[cmake] CMake based buildsystem

fetzerch opened the pull request:
This is the current state of the CMake based buildsystem I've been working on the past weeks.
Since it is now possible to build the Kodi executable on Linux, Windows and OSX, @FernetMenta asked me to create a pull request for it so that we can gather feedback. I've tried to document the usage in https://github.com/fetzerch/xbmc/blob/cmake-master/project/cmake/README.md as good as possible, but don't hesitate to ask if you run into problems.

It would be really great to get some feedback on the generated Visual Studio and Xcode projects from devs that are familiar with these IDEs. The CMake project builds smaller static libraries per subfolder in xbmc hence the IDE projects look a bit different to what you might be used to.  

The current state as well as open todos can be found in this forum thread: http://forum.kodi.tv/showthread.php?tid=255893

And most importantly: Kudos to @notspiff for starting this.

List of open points:
- [x] C++11 check
- [x] win32: Generate pdb in release build
- [x] osx: CMake error (dependencies for) TexturePacker
- [x] kodi binary name
- [x] test execution with make check (on linux)

fetzerch committed:
[win32/PlatformDefs.h] Compatibility with VS2013

Starting with Visual Studio 2013 (_MSC_VER 1800) there is built-in support
for va_copy and strtoll. Therefore the compatibility macros only need to
be defined for older versions.

This commit fixes warnings/errors that occur when headers are included in
a different order:

- Including PlatformDefs.h before stdarg.h causes:
  C4005: 'va_copy': macro redefinition
- Including PlatformDefs.h before stdlib.h causes:
  C2039: 'strtoll': is not a member of '`global namespace`'
  C2873: 'strtoll': symbol cannot be used in using-declaration

fetzerch committed:
[EventClient]: Fix ambigious symbol error

Fixes warning with Visual Studio 2013:

EventClient.cpp(240) : error C2872: 'PT_MOUSE' : ambiguous symbol
  could be 'C:\Program Files (x86)\Windows Kits\8.1\include\um\winuser.h(5980)
    : tagPOINTER_INPUT_TYPE PT_MOUSE'
  or 'xbmc\network\EventPacket.h(150): EVENTPACKET::PacketType PT_MOUSE'

fetzerch committed:
[EndianSwap] Fix include config.h in EndianSwap.h

Inclusion of config.h is usually guarded by '#if (defined HAVE_CONFIG_H) && (!defined TARGET_WINDOWS)'.
In EndianSwap.h it is not which can lead to inclusion of any config.h found in the include directories.

fetzerch committed:
[cmake] Add generated files to gitignore

fetzerch committed:
[cmake] Document CMake based buildsystem in README.md

fetzerch committed:
[cmake] Add define BUILDING_WITH_CMAKE for Windows

This adds a define BUILDING_WITH_CMAKE that allows to differentiate the
build system and have CMake specific adaptions in the code.
This allows to:

- Use the CMake options to conditionally enable/disable features.
- Get rid of the #pragma comment(lib *) directives and properly link
  found libraries.

This define is just there for the transitional phase. Once the VS
project is dropped the define gives a good indication where legacy code
can be removed.

Additionally remove 2 unused defines from system.h:
HAVE_LIBMP3LAME, HAVE_LIBVORBISENC

fetzerch committed:
[cmake] Adapt Toolchain.cmake.in for CMake based buildsystem

fetzerch committed:
[cmake] Add CMakeLists.txt to xbmc subfolders

fetzerch committed:
[cmake] Add CMake files to tools/depends

Add CMakeLists.txt and find modules for dependencies.

- OSXRemote
- JsonSchemaBuilder
- TexturePacker
- libsquish
- crossguid
- ffmpeg

fetzerch committed:
[cmake] Add CMake files to lib

Add CMakeLists.txt for the lib directory.

- lib/addons
- UnrarXLib
- libUPnP
- libexif

fetzerch committed:
[cmake] Add main CMake file

fetzerch committed:
[cmake] Add FindXXX.cmake modules

Add CMake find modules for finding dependencies.

fetzerch committed:
[cmake] Add treedata

Configures subdirectories used for the different platforms.

fetzerch committed:
[cmake] Add platform configuration scripts

fetzerch committed:
[cmake] Add installation data configuration

Paxxi commented:
haven't looked at it yet, will have a look at the project files it creates.

I just want to say AWESOME WORK! both @fetzerch and @notspiff this has been a long time coming and it's nice that someone want to undertake the crazy amount of work that it is.

My initial feeling is that unless there's some critical issues we want this merged as soon as possible so it gets more visibility and testing. We can wait with removing the xcode/vsprojects until we feel this is "done" and then get rid of them.

FernetMenta commented:
awesome!

wsnipex commented on project/cmake/modules/FindCXX11.cmake:
```diff
@@ -0,0 +1,18 @@
+if(CXX_FLAG_CXX11)
+  add_options (CXX ALL_BUILDS "-std=c++11")
+  set(CXX_STD11_FLAGS "-std=c++11")
+else()
```
we have to error out here, since we require c++11

fetzerch replied:
Thanks for pointing that out. There are a few ways to do that in CMake. Do we have minimum compiler versions defined somewhere? Because even if a compiler has the `-std=c++11` flag, it doesn't mean that it support all C++11 features that are used in the project. And Microsoft compilers don't have the flag at all. There's also a way to specify needed C++ features and CMake figures out the options (https://cmake.org/cmake/help/latest/manual/cmake-compile-features.7.html). But that might be a bigger task considering the project size.

notspiff replied:
find_package_handle_standard_args handles the error message and shit.

wsnipex replied:
gcc >=4.7.2 is known to work, 4.6 doesn't.
not sure about clang, but I guess 3.6 works
and VS2013 afaik

notspiff replied:
then remove the c++0x block from the find rule. 4.6 doesn't announce c++11.

fetzerch replied:
Starting with CMake 3.1 we could simply use this in the main CMakeLists.txt which works [cross platform and for all interesting compilers](https://cmake.org/cmake/help/latest/manual/cmake-compile-features.7.html#supported-compilers). If there are no objections, I'd go for this. 
```
set(CMAKE_CXX_STANDARD 11)
set(CMAKE_CXX_STANDARD_REQUIRED ON)
set(CMAKE_CXX_EXTENSIONS OFF)  # Strict standard, no compiler specific extensions (such as gnu11)
```

afedchin commented:
how to debug Kodi on Windows with this?

Paxxi commented:
@afedchin you run it like "cmake -G "Visual Studio 12" and it'll create a visual studio solution which lets you work as normal. It will look different but it should be perfectly usable

afedchin commented:
@Paxxi I don't think so. for binary addons we have tools/windows/prepare-binary-addons-dev.bat and it looks not same simple as your line. 

@fetzerch please create a bat tool for debug in VS

Voyager1 commented:
@afedchin my assumption is that BuildSetup.bat still has to be modified to make the cmake calls. For release build it will use NMAKE, and for debugging purposes we need another .bat file to create / update the Studio solution files...

AchimTuran commented:
@Voyager1

> For release build it will use NMAKE
Why not msbuild?

@fetzerch
Thx fr creating this PR and others that where involved in this project. This will give us better posibilities to integrate the binary add-on builds.

I will try to use it in the next days ;-)

fetzerch commented:
@afedchin I'll create a bat file similar to the one you mentioned. But @Paxxi is correct: it's as simple as executing `cmake -G "Visual Studio 12 (2013)" ...` (see readme linked in the first post). Binary addons are a bit more complex because they have a few required parameters.

What changes with CMake is that the generated VS solution is "read only". Compiler settings, removing or adding files and libraries has to be done in the CMake files. With the big advantage, that it's centralized and doesn't have to be done for 3 different build systems like it is today.

@AchimTuran: The commands to build with NMake or MSBuild are very similar. From my experience so far, NMake or more specifically [jom.exe](https://wiki.qt.io/Jom) does a much better job in parallelizing compared to MSBuild and can speed up the build a lot (on my machine it takes only half of the time). This would make it a perfect candidate for CI builds where execution time matters.

FernetMenta commented:
On OSX I get this error:

```
-- Found Threads: TRUE  
-- Found Lzo2: /Users/Shared/xbmc-depends/macosx10.11_x86_64-target/include/lzo  
-- Could NOT find ZLIB (missing:  ZLIB_LIBRARY) (found version "1.2.5")
CMake Error at /usr/local/Cellar/cmake/3.2.1/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:138 (message):
  Could NOT find PNG (missing: PNG_LIBRARY PNG_PNG_INCLUDE_DIR)
Call Stack (most recent call first):
  /usr/local/Cellar/cmake/3.2.1/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:374 (_FPHSA_FAILURE_MESSAGE)
  /usr/local/Cellar/cmake/3.2.1/share/cmake/Modules/FindPNG.cmake:121 (find_package_handle_standard_args)
  /Users/sdf/progs/src/xbmc/tools/depends/native/TexturePacker/CMakeLists.txt:4 (find_package)
```

Razzeee commented on project/cmake/README.md:
````diff
@@ -0,0 +1,131 @@
+### Windows with Visual Studio project files
+
+```
+cmake -G "Visual Studio 12 (2013)" <KODI_SRC>/project/cmake/
````
It seems to be "Visual Studio 12 2013"

ace20022 replied:
it is ```cmake -G "Visual Studio 12"```, at least I use that for libbluray ans libudfread ;)

Razzeee replied:
Both work

Razzeee commented on project/cmake/CMakeLists.txt:
```diff
@@ -0,0 +1,351 @@
+# Required dependencies
+set(required_deps Sqlite3 FreeType PCRE Cpluff
+                  TinyXML Python Yajl Xslt
+                  JPEG Lzo2 Fribidi TagLib Boost FFMPEG CrossGUID)
```
boost needs to go

Razzeee commented:
This is what it looks like in visual studio, not sure everything compiled right.
![image](https://cloud.githubusercontent.com/assets/5943908/12881144/9b66b08c-ce47-11e5-812d-500db76754a1.png)

Razzeee commented on xbmc/guilib/CMakeLists.txt:
```diff
@@ -0,0 +1,246 @@
+                      MatrixGLES.cpp
+                      TextureGL.cpp
+                      TexturePi.cpp)
+  list(APPEND SOURCES GUIFontTTFGL.h
```
Should probably be APPEND HEADERS?

afedchin commented:
1. I was not be able to build Kodi with generated solution
```
3>  FX compile D:/xbmc/system/shaders/guishader_vert.hlsl
3>  'FXC-NOTFOUND' is not recognized as an internal or external command,
3>  operable program or batch file.
```
2. kodi.bin project not well configured for debugging.
3. The generated solution is not usable for me (just for me)

Razzeee commented:
See https://github.com/xbmc/xbmc/pull/9083#discussion_r52144115 that fixes it

afedchin commented:
@Razzeee thanks, it solves the error. but resulting exe looks strange. After debugging is started it opens a console window and Kodi window (yes two separate windows) and then after splash immediately close all windows without any reason.

If I start Kodi from console I get this
```
D:\xbmc\project\Win32BuildSetup\BUILD_WIN32>kodi.bin.exe -p
ImportError: No module named site
``` 
looks like a python error.

afedchin commented:
so, cmake creates kodi.bin project as console project, to solve this need WIN32 modifier for add_executable
```
add_executable(${APP_NAME_LC}.bin WIN32 ...
```
take a look https://github.com/afedchin/xbmc/commit/0da607393d8fd2f5b1086001d9b3c229bb873b41

but there is an unclear error with python. It cannot to be initialized via Py_Initialize() and force closes the kodi application.

afedchin committed:
[win32] fix manifest to properly load msvcr90.dll required for python before it will try to initialize.

afedchin commented:
here is a hackish fix for python https://github.com/afedchin/xbmc/commit/9e6ae56577d0ba5a95b492ada6ca39632dd6d4ef

afedchin commented:
one more thing for windows, in release need to generate .pdb also

Razzeee commented:
btw how to run the test suite now?

akva2 commented:
make check / check target in project.

fetzerch commented:
Wow, thanks for your comments and the patches! Nice to see that this got so much attention.
I've added some fixes to the PR. Unfortunately I couldn't check them them on Windows today, but I'll try to do that the next days.

Still missing:
- C++11 check
- Generate pdb in release build
- CMake error on OSX. @FernetMenta: I'll try to reproduce it as soon as I can. Just to be sure that it's not something with the toolchain file. Could you try fully cleaning the build-dir and then run cmake with an absolute path to the toolchain file?

@akva2: I've tried `make check` today on Linux, but building kodi-test failed because of multiple definition of main. It's defined once in xbmc-test.cpp and once in platform_posix. I'd guess it's best fixed by adding a NO_TEST_DEPENDS argument to core_add_library and store the libs needed for the tests in a separate variable. Strange that this pops up just now. I'm almost sure I tried that some weeks ago and it worked.

akva2 commented:
yes, the platform stuff got introduced not too long back.

the fix in my fork, which is see i haven't pushed from my machine in the office, is to not build main.o into the posix_platform library, and then add is explicitly to the kodi.bin target. obviously you'd have to do some abstraction / if spaghetti somewhere. but you can limit it to that single object.

akva2 commented:
basically what you undid here https://github.com/FernetMenta/xbmc/commit/1446d20042c88de3f96765c4b0889787049fdafa except with modified paths..

FernetMenta commented:
@fetzerch 
> Could you try fully cleaning the build-dir and then run cmake with an absolute path to the toolchain file?

This is what I did. I created new new folder "build": <KODI_SRC>/build Then used absolute path names.

afedchin commented:
@fetzerch thanks for fix.
1. Can we change win32 executable name? For non windows platforms it is kodi.bin, but for windows it will be a kodi.bin.exe. I think it should be simply kodi.exe
2. here is a little fix generate .pdb in release https://github.com/afedchin/xbmc/commit/687038c089607c41eb8cb68ca4a19e211f5ac40a

Montellese commented:
IIRC we should not set the project name to "kodi.bin" but set it to "kodi" and for every platform use `set_target_properties` to set `SUFFIX` to `bin` or `exe` or whatever.

akva2 commented:
^^

afedchin committed:
[SystemGlobals] make sure what CSectionLoader instance still alive before destruct any global variable which extends from DllDynamic (as an example g_curlInstance).

There is a potential problem of accessing the object after it has been destroyed. This only happens if you access the object from the destructor of another global variable (by global, I am referring to any non-local static variable).

The solution is to make sure that you force the order of destruction. The order of destruction is the exact inverse of the order of construction. So if you access the object in your destructor, you must guarantee that the object has not been destroyed. To do this, you must just guarantee that the object is fully constructed before the calling object is constructed.

afedchin commented:
testes a little more. another fix is here https://github.com/afedchin/xbmc/commit/ce3eaedea6ca3b0c974c3099d12e8f12e1840e74 commit message describes a problem which occurs on windows during testing. @FernetMenta take a look also please.

notspiff commented:
hmm. not that i'm opposed to switching to a vanilla singleton, but this looks like the mess XBMC_GLOBALS is supposed to handle. why does it differ in this build i wonder.

afedchin commented:
I don't know exactly why it differs with cmake build but I think what main difference in that the DllLibCurlGlobal and CSectionLoader are in different libraries (different projects), whereas in the classic sln these classes in same project.

notspiff commented:
that makes sense. alternative solution is probably to reorder the libs then but that could get messy.

FernetMenta commented:
@afedchin I would avoid static members that are the reason why we have something like XMBC_GLOBALS. IMO it is better to instantiate the object on first access. Then you avoid placement of the variables into .BSS or .DATA

afedchin commented:
i've updated the fix https://github.com/afedchin/xbmc/commit/02e7e50637e41591678281ac1eebf4ee32d4b7c8. but I think we need a proper complex fix to avoid those situations in the future.

fetzerch commented:
Next round of fixes:
- OSX cmake error should be fixed now. @FernetMenta: please confirm
- Binary name is now kodi.exe on Windows, kodi.bin for other platforms
- Test execution should work now (with `make check`)

FernetMenta commented:
@fetzerch it passed the last issue but I got a new one:

```
-- Found Threads: TRUE  
CMake Error at /usr/local/Cellar/cmake/3.2.1/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:138 (message):
  Could NOT find Lzo2 (missing: LZO2_INCLUDE_DIRS LZO2_LIBRARIES)
Call Stack (most recent call first):
  /usr/local/Cellar/cmake/3.2.1/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:374 (_FPHSA_FAILURE_MESSAGE)
```

FernetMenta commented:
This does not have to be perfect in the first round, right? I recommend to merge this so it can get a broader user base.

wsnipex commented:
+1
since we don't depend on it yet, it has time to mature and this works best with broader usage.

fetzerch committed:
[win32] Initialize converters before usage

fetzerch commented:
Ok. I've squashed the fixes into the original commits and rebased the PR (AddonBuilder, PVRRecordingsPath got added in the meantime) so that it builds again with the current master.

hudokkow commented on project/cmake/scripts/linux/macros.cmake:
```diff
@@ -0,0 +1,73 @@
+    set(${lib}_SONAME ${${lib}_SONAME} PARENT_SCOPE)
+  endif()
+  if(arg_REQUIRED AND NOT ${lib}_SONAME)
+    message(FATAL_ERROR "Could not find dynamically loadable library ${LIB}")
```
... ${lib} ?
Otherwise output is a bit ghostly... ;-)

hudokkow commented on project/cmake/scripts/darwin/macros.cmake:
```diff
@@ -0,0 +1,108 @@
+    endif()
+  endforeach()
+  if(arg_REQUIRED AND NOT ${lib}_SONAME)
+    message(FATAL_ERROR "Could not find dynamically loadable library ${LIB}")
```
Same here.

FernetMenta commented:
jenkins build this please

fetzerch commented:
@FernetMenta: I've fixed these 2 things that @hudokkow mentioned. Is it ok to force push while jenkins is running?

Razzeee commented:
@fetzerch yes, but it will abort the build

MartijnKaijser commented:
no it doesn't abort the build on jenkins itself. it just won't show status here

FernetMenta commented:
@fetzerch have you completed your changes? should we start jenkins again?

hudokkow commented:
@fetzerch, a few things I noticed:

> -- Found Doxygen: /usr/bin/doxygen (found version "1.8.9.1") 
-- Warning: No doxygen installed. The Api will not have any docstrings.
CMake Error at scripts/linux/install.cmake:22 (install):
  install TARGETS given target "kodi.bin" which does not exist in this
  directory.
Call Stack (most recent call first):
  CMakeLists.txt:338 (include)

I do have doxygen installed. Config output doesn't show anything weird except
> -- VAAPI enabled: No

but libva* packages are present.

BTW, should <code>Generator: Single-configuration: Release (Unix Makefiles)</code> be the default? I thought you went with Debug for default.

akva2 commented:
your vaapi packages are outdated. you need things newer than any distro ships.

hudokkow commented:
Nope, that's not it.
./configure doesn't complain and shows <code>VAAPI:	Yes</code>

akva2 commented:
that's because configure blindly accepts any version.

anyways, things are more messy it seems. it used to be va.pc, now it's libva.pc. furthermore the one in my repo is has pkg-config version set to 0.35.0, while the library is 1.3.0. stupid crappy ubuntu packages. this has confused me for a long time.

akva2 commented:
right, libva has api and abi versioning and the pc file follows the abi. e.g. 1.6.0 has pkg-config version 0.38. so the check for 1.4.0 is wrong.

hudokkow commented:
Just checked:
> libva-dev:
  Installed: 1.6.1~wily
  Candidate: 1.6.1~wily
  Version table:
 *** 1.6.1~wily 0
        500 http://ppa.launchpad.net/wsnipex/vaapi/ubuntu/ wily/main amd64 Packages
        100 /var/lib/dpkg/status
     1.6.0-1 0
        500 http://pt.archive.ubuntu.com/ubuntu/ wily/universe amd64 Packages

> libva1:
  Installed: 1.6.1~wily
  Candidate: 1.6.1~wily
  Version table:
 *** 1.6.1~wily 0
        500 http://ppa.launchpad.net/wsnipex/vaapi/ubuntu/ wily/main amd64 Packages
        100 /var/lib/dpkg/status
     1.6.0-1 0
        500 http://pt.archive.ubuntu.com/ubuntu/ wily/universe amd64 Packages

akva2 commented:
yes, and now see your pc file - it will say 0.38.0!

hudokkow commented:
Heheh, your right!

akva2 commented:
doh, this is even more awesome. "somebody" has used  pkgconfig syntax for find_library.

wsnipex commented:
hrhr :)
findvaapi and findvdpau should probably use pkg_config instead. Imho no need for a find_library fallback.

akva2 commented:
the fallbacks are there as a general policy to handle depends shit. i got fed up by the pc being patched in by ubuntu packaging in too many libs for me to bother. (yes might not apply here). i will add pkgconfig for vdpau.

notspiff committed:
fixed: tear down settings environment in xbmc-test

somehow things break if this is done in the global's destructors.

makes test-execution almost deterministic

wsnipex commented:
if I'm not blind, vaapi doesn't use pkconfig either.

akva2 commented:
that was the problem. somebody thought it did and appended >= 1.4.0 on  a find_library call.

notspiff commented:
https://github.com/fetzerch/xbmc/pull/1

hudokkow commented:
*Somebody* solved the problem, heheh. Building now. Thanks Arne!

fetzerch commented:
Added fixes by @notspiff. Anything else for this initial version?

Paxxi commented:
Just hit the button :)

MartijnKaijser commented:
jenkins build this please

MartijnKaijser commented:
it has the usual win32 addon build failure.
